### PR TITLE
Add Custom Builds : fixes #29

### LINKED ISSUE #29: Should have - "Custom" template type
## Expected Behaviour

It should be possible to "build" a function which isn't using one of the known / supported template types. An example would be the shrink_image example in the repo.

Snippet:

```yaml
  shrink-image:
    lang: custom
    handler: ./samples/image-magick
    image: functions/resizer
    fprocess: "convert - -resize 50% fd:1"
```

The context is whatever is in the `handler` folder and the Dockerfile in the root of the directory will be used for the build.

## Current Behaviour

Build manually and remove the `lang` element then mark the item as `skip_build`

## Possible Solution

When setting the language to `custom` then a `docker build` is invoked in the handler folder directly.

## Context

This will add versatility - especially for languages which don't suit dynamic templating like Golang.

### PR DESCRIPTION
**Pull request by dudymas**
## Description

The builder handles a new language type, specified by the string 'custom'
This type treats the 'handler' as a directory with a dockerfile and builds it.

## Motivation and Context

This is for issue #29 

## How Has This Been Tested?

This code has been manually tested using rebuilt faas-cli binaries on a windows machine.

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the code style of this project.
(note: I spaced out some string concats in the builder. Holler if I shouldn't do that)
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/alexellis/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by dudymas** [comment]
I don't mind squashing this. I've left commits separated to expedite review and rework

File: README.md [hunk 1]
```diff
@@ -98,6 +98,24 @@ $ uname -a | curl http://localhost:8080/function/nodejs-echo--data-binary @-
 
 > For further instructions on the manual CLI flags (without using a YAML file) read [manual_cli.md](https://github.com/alexellis/faas-cli/blob/master/MANUAL_CLI.md)
 
+#### Adding custom language support
+
+If you want to add a function that does not have a template (such as `go`),
+you only need to place a dockerfile with in a directory such that:
+
+1. The dockerfile builds successfully with appropriate settings:
```
Comment (alexellis) [inline README.md]: dockerfile -> Dockerfile

File: README.md [hunk 2]
```diff
@@ -98,6 +98,24 @@ $ uname -a | curl http://localhost:8080/function/nodejs-echo--data-binary @-
 
 > For further instructions on the manual CLI flags (without using a YAML file) read [manual_cli.md](https://github.com/alexellis/faas-cli/blob/master/MANUAL_CLI.md)
 
+#### Adding custom language support
```
Comment (alexellis) [inline README.md]: Not just custom language, this could be a pre-existing binary. For instance take the samples here -> https://github.com/alexellis/faas/tree/master/sample-functions

and here -> https://github.com/alexellis/faas/blob/master/docker-compose.yml

File: sample/golang-echo/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,12 @@
+FROM golang:1.9-alpine
```
Comment (alexellis) [inline sample/golang-echo/Dockerfile]: What's the current LTS version?

Comment (dudymas) [inline sample/golang-echo/Dockerfile]: 3.5 ... setting that nowish

Comment (dudymas) [inline sample/golang-echo/Dockerfile]: of course, all the older stable container tags are unsupported AND vulnerable. For now, I'm setting this to `golang:1.9-alpine3.6`
would that be okay. If not, got a preferred golang container for this sample? Or I could just scrap it for fsharp instead

Comment (alexellis) [inline sample/golang-echo/Dockerfile]: Docker have official Golang images, are you certain that the LTS Golang version is vulnerable? Can you link to something showing that?

File: README.md (hunk 2 above)
Comment (dudymas) [inline README.md]: I updated the wording of this to:
> Adding custom function support

will that work?

File: sample/golang-echo/Dockerfile (hunk 3 above)
Comment (dudymas) [inline sample/golang-echo/Dockerfile]: https://hub.docker.com/r/library/golang/tags/ only lists the alpine3.6 versions as good. And I misunderstood what you meant by LTS. I thought you were talking about which version of alpine has best support.

Comment (alexellis) [inline sample/golang-echo/Dockerfile]: I meant LTS of Golang. By the way - where are you seeing the bad flags on the images? How about 1.8-alpine3.6?

Comment (alexellis) [inline sample/golang-echo/Dockerfile]: Here's a different idea - can we do the sample in a different PR? .. just the custom support in this one and then we can get it merged.

Comment (dudymas) [inline sample/golang-echo/Dockerfile]: that's a solid idea... picking this fellow out now

**Comment by alexellis** [comment]
Hi @dudymas,

Please can you correct the case of dockerfile->Dockerfile?

Can you show the output from the build action with the a Golang sample like this one?

https://github.com/alexellis/faas/tree/master/sample-functions/DockerHubStats

At the same time would you like to update that sample to use a multi-stage build? (it will need ca-certs / ca-certificates as a dependency for the web scraping)

File: README.md [hunk 4]
```diff
@@ -98,6 +98,24 @@ $ uname -a | curl http://localhost:8080/function/nodejs-echo--data-binary @-
...
+you only need to place a dockerfile within a directory such that:
+
+1. The dockerfile builds successfully with appropriate settings:
+   * The `fprocess` env var should be the path + flags for an executable binary
+   * The binary should accept stdin and print to stdout
+2. You specify the directory that contains the dockerfile as a 'handler':
+   * For the yaml, just put in the path. You can see this in `samples.yml`
+  for the golang-echo function
+   * On the command line, use the `-handler` flag
+3. You specify the the `language` as `custom`:
```
Comment (johnmccabe) [inline README.md]: PR is looking good @dudymas, the only concern I have is that `custom` is a bit too broad for what is in effect a Dockerfile specific template.

Changing the language from `custom` to `dockerfile` would make its behaviour unambiguous.

**Comment by alexellis** [comment]
@dudymas thanks for working on this. I've added some changes in PR https://github.com/alexellis/faas-cli/pull/52 which mean we can get this into the code faster. I made a couple of changes: Dockerfile is the language instead of custom and there is no temporary build folder in use (since there is no template).

**Comment by dudymas** [comment]
No worries. I had a busy week, and I know how fast this PR was put together ;-)
